python/pyiceberg/io/fsspec.py

```diff
@@ -15,28 +15,76 @@
 # specific language governing permissions and limitations
 # under the License.
 """FileIO implementation for reading and writing table files that uses fsspec compatible filesystems"""
-
-from functools import lru_cache
-from typing import Callable, Union
+import logging
+from functools import lru_cache, partial
+from typing import Callable, Dict, Union
 from urllib.parse import urlparse
 
+import requests
+from botocore import UNSIGNED
+from botocore.awsrequest import AWSRequest
 from fsspec import AbstractFileSystem
+from requests import HTTPError
 from s3fs import S3FileSystem
 
+from pyiceberg.exceptions import SignError
 from pyiceberg.io import FileIO, InputFile, OutputFile
 from pyiceberg.typedef import Properties
 
+logger = logging.getLogger(__name__)
+
+
+def s3v4_rest_signer(properties: Properties, request: AWSRequest, **_) -> AWSRequest:
+    signer_url = properties["uri"].rstrip("/")
+    signer_headers = {"Authorization": f"Bearer {properties['token']}"}
+    signer_body = {
+        "method": request.method,
+        "region": request.context["client_region"],
+        "uri": request.url,
+        "headers": {key: [val] for key, val in request.headers.items()},
+    }
+    try:
+        response = requests.post(f"{signer_url}/v1/aws/s3/sign", headers=signer_headers, json=signer_body)
+        response.raise_for_status()
+        response_json = response.json()
+    except HTTPError as e:
+        raise SignError(f"Failed to sign request: {signer_headers}") from e
+
+    for key, value in response_json["headers"].items():
+        request.headers.add_header(key, value[0])
+
+    return AWSRequest
+
+
+SIGNERS: Dict[str, Callable[[Properties, AWSRequest], AWSRequest]] = {"S3V4RestSigner": s3v4_rest_signer}
+
 
 def _s3(properties: Properties) -> AbstractFileSystem:
     client_kwargs = {
         "endpoint_url": properties.get("s3.endpoint"),
         "aws_access_key_id": properties.get("s3.access-key-id"),
         "aws_secret_access_key": properties.get("s3.secret-access-key"),
     }
+    config_kwargs = {}
+    register_events: Dict[str, Callable] = {}
+
+    if signer := properties.get("s3.signer"):
+        logger.info("Loading signer %s", signer)
+        if singer_func := SIGNERS.get(signer):
+            singer_func_with_properties = partial(singer_func, properties)
+            register_events["request-created.s3"] = singer_func_with_properties
```

Comment:
Is this using `request-created.s3` rather than `before-sign.s3` to avoid the error that you were talking about? `before-sign` seems like a safer one to use to me.

Reply:
I just checked, and the `before-sign.s3` works just as well. The error that I was describing is unrelated to this, it seems that the signer can be registered multiple times, and then the request gets signed more than once, resulting in a 403.

Reply:
I think it would be best to solve that in a separate PR, since that I would also expect that to affect the caching of the filesystem